refactor: Introduce "tracing" plugin and switch request pipeline to use it.

abernix opened the pull request:
This commit introduces a `plugin` (named) export from the `apollo-tracing` package, which uses the new "plugin" hooks rather than the previous implementation (exported as `TracingExtension`) which uses the soon-to-be-deprecated "extensions" hooks.  The functionality is intended to be identical, in spirit.

Since the delta of the commits was otherwise confusing, I've left the `TracingExtension` present and exported and will remove it in a subsequent commit.

Briefly summarizing what the necessary changes were:

1. We no longer use a class instance to house the extension, which was necessitated by the `graphql-extensions` API.  This means that uses of `this` have been replaced with function scoped variables by the same name.
2. The logic which actually does the formatting (previously handled by the `format` method in `graphql-extension`, now takes place within the plugin API's `willSendResponse` method.

Finally, this removes the rendered-unused/unnecessary `TracingExtension` extension after providing its replacement.

trevor-scheer commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -783,11 +784,13 @@ export class ApolloServerBase {
   private ensurePluginInstantiation(plugins?: PluginDefinition[]): void {
-    if (!plugins || !plugins.length) {
-      return;
+    const pluginsToInit = [...plugins || []];
```
Would a default argument `[]` have the desired effect?

abernix replied:
It would!  However, in reality, this code changes in an upcoming PR, so I'm going to leave it as is here — as valid as a suggestion as this is. 😄

abernix replied:
To make this more clear: I ended up incorporating this feedback in a preparatory way in the branch that this PR merges into (https://github.com/apollographql/apollo-server/pull/3988/commits/4b59b02b0da2602ea1f0bcb399687ad170699e07) rather than trying to bring it back later.  Good move, I think.  Thanks for noting this.

trevor-scheer commented on packages/apollo-tracing/src/index.ts:
```diff
@@ -33,6 +33,97 @@ interface ResolverCall {
+    let duration: HighResolutionTime | undefined;
+    const resolverCalls: ResolverCall[] = [];
+
+    startWallTime = new Date();
+    startHrTime = process.hrtime();
```
Is tracing intended to include timing for parse/validate? I would've imagined these happening within `executionDidStart`.

abernix replied:
Neither tracing or engine reporting really highlight the parsing and validation phases.  I suspect there is some outstanding intention to provide that insight eventually and that those would leverage this `startHrTime` as offsets to their timings, much in the way that this `startHrTime` is used as offsets for individual resolver timings right now.

For now, this maintains identical behavior to the existing behavior so we're not bringing any breaking changes along with this critical-to-unblock-3.x re-factor.  (We need to be careful to

trevor-scheer commented on packages/apollo-tracing/src/index.ts:
```diff
@@ -33,6 +33,97 @@ interface ResolverCall {
+          response.extensions || (response.extensions = Object.create(null));
+
+        if (typeof extensions.tracing !== 'undefined') {
+          throw new Error("The tracing information already existed.");
```
Possibly by another plugin? Curious what we're defending against here.

abernix replied:
Hopefully more clear with: https://github.com/apollographql/apollo-server/pull/3991/commits/7e4cc4b7d99aeaffe6a27b2cf97b98d7713c2a5c

trevor-scheer commented on packages/apollo-tracing/src/index.ts:
```diff
@@ -33,6 +33,97 @@ interface ResolverCall {
+                ? durationHrTimeToNanos(resolverCall.endOffset) - startOffset
+                : 0;
+              return {
+                path: [...responsePathAsArray(resolverCall.path)],
```
Any reason to return a different array here than what the function returns?

A fix in case this wasn't the intent:
```suggestion
                path: responsePathAsArray(resolverCall.path),
```

abernix replied:
This maintains the behavior from the original implementation and I didn't second-guess it.  I believe the `responsePathAsArray` (also, `graphql-js`'s `pathToArray`) returns a linked list so I'm assuming this was meant to clone the array (as the spread notation would do in this case).

abernix replied:
This is a reasonable thing to scrutinize, and I wish that a comment indicated the reason for this when it was introduced.  This code was originally imported from a now-defunct `apollo-cache-control-js` package (which I happen to still have cloned locally, but doesn't provide any more insight).  In the previous implementation, it was not using this technique, but it appears that this spread was added to _get tests passing_ in https://github.com/apollographql/apollo-server/commit/3523e4ae02ce6ed2c8c211f141bc7f87ee42474e within #1294.

Anyhow, since the intent was largely to maintain identical behavior to what existed before.  I'd rather not jeopardize the success of landing this large refactor based on further changing implementation details and instead just make the change to the integration surface (plugin vs. extension).

glasser commented:
While this work (whose review I'll leave up to @trevor-scheer) is worthwhile, can you while you're at it (and as part of describing it in CHANGELOG) update its description/docs to make it clear that it's more or less deprecated? My understanding is that the primary use case of apollo-tracing was for AGM but that only worked via engineproxy; the README does not make it clear that that's a highly deprecated use case. I think there's also some graphql-playground support for it too.  I think users get confused a lot into thinking this package (and its enabling flag) has anything to do with modern AGM reporting, but it doesn't.

abernix commented on CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ The version headers in this history reflect the versions of Apollo Server itself
 > The changes noted within this `vNEXT` section have not been released yet.  New PRs and commits which introduce changes should include an entry in this `vNEXT` section as part of their development.  When a release is being prepared, a new header will be (manually) created below and the the appropriate changes within that release will be moved into the new section.
 
 - `apollo-server-lambda`: Support file uploads on AWS Lambda [Issue #1419](https://github.com/apollographql/apollo-server/issues/1419) [Issue #1703](https://github.com/apollographql/apollo-server/issues/1703) [PR #3926](https://github.com/apollographql/apollo-server/pull/3926)
+- `apollo-tracing`: This package's internal integration with Apollo Server has been switched from using the soon-to-be-deprecated`graphql-extensions` API to using [the request pipeline plugin API](https://www.apollographql.com/docs/apollo-server/integrations/plugins/).  Behavior should remain otherwise the same.  [PR #3991](https://github.com/apollographql/apollo-server/pull/3991)
```
```suggestion
- `apollo-tracing`: This package is considered deprecated and — along with its `tracing: Boolean` configuration option on the `ApolloServer` constructor options — will cease to exist in Apollo Server 3.x.  Until that occurs, we've updated the _internal_ integration of this package with Apollo Server itself to use the newer [request pipeline plugin API](https://www.apollographql.com/docs/apollo-server/integrations/plugins/), rather than the _also_ soon-to-be-deprecated-`graphql-extensions` API it previously leveraged.  The behavior should remain otherwise the same.  [PR #3991](https://github.com/apollographql/apollo-server/pull/3991)
```

abernix commented:
I have reverted (https://github.com/apollographql/apollo-server/pull/3991/commits/ef4ed58cb5e14bb638e4580760673de347a88de0) the comment I added to the `CHANGELOG.md` (https://github.com/apollographql/apollo-server/pull/3991/commits/98bae446d37704939bd7e0bf010503c66215dfd8) in response to @glasser's https://github.com/apollographql/apollo-server/pull/3991#issuecomment-614796233 about _more or less_ deprecation of this concept since it is, in fact, still very much _more or less_.

While I agree with us framing it appropriately and clearly, let's do that _ASAP_ outside the scope of this PR.
